raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,697 @@
+            channel_identifier,
+            partner_address,
+            nonce,
+            locksroot,
+            transferred_amount
+        );
+
+        ChannelClosed(channel_identifier, msg.sender);
+    }
+
+    /// @notice Called on a closed channel, the function allows the non-closing participant
+    /// to provide the last balance proof, which modifies the closing participant's state.
+    /// Can be called multiple times, only by the non-closing participant.
+    /// @param channel_identifier The channel identifier - mapping key used for `channels`.
+    /// @param nonce Strictly monotonic value used to order transfers.
+    /// @param transferred_amount Total amount of tokens transferred by the channel partner
+    /// to the channel participant who calls the function.
+    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
+    /// @param additional_hash Computed from the message. Used for message authentication.
+    /// @param closing_signature Signature of the closing participant on the balance proof data.
+    function updateTransfer(
+        uint256 channel_identifier,
+        uint64 nonce,
+        uint256 transferred_amount,
+        bytes32 locksroot,
+        bytes32 additional_hash,
+        bytes closing_signature)
+        public
+    {
+        // The caller has to be a channel participant
+        require(channels[channel_identifier].participants[msg.sender].initialized);
+
+        // The closer is not allowed to call updateTransfer
+        require(closing_requests[channel_identifier].closing_participant != msg.sender);
+
+        // Call the private function for the actual logic and constraints of updateTransfer
+        updateTransferPrivate(
+            channel_identifier,
+            nonce,
+            transferred_amount,
+            locksroot,
+            additional_hash,
+            closing_signature
+        );
+    }
+
+    /// @notice Called on a closed channel, the function allows the non-closing participant to
+    // provide the last balance proof, which modifies the closing participant's state. Can be
+    // called multiple times by anyone, as long as they provide signatures from both participants.
+    /// @param channel_identifier The channel identifier - mapping key used for `channels`.
+    /// @param nonce Strictly monotonic value used to order transfers.
+    /// @param transferred_amount Total amount of tokens transferred by the channel partner
+    /// to the channel participant who calls the function.
+    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
+    /// @param additional_hash Computed from the message. Used for message authentication.
+    /// @param closing_signature Closing participant's signature of the balance proof data.
+    /// @param non_closing_signature Non-closing participant signature of the balance proof data.
+    function updateTransferDelegate(
+        uint256 channel_identifier,
```

Comment:
the token_address is missing here too

Reply:
Same. 
I will add the `chain_id`.
I used the TokenNetwork contract address in `recoverAddressFromSignature`.